File: src/main/java/com/github/jactor/persistence/aop/ModifierAspect.java
package com.github.jactor.persistence.aop;

import com.github.jactor.persistence.entity.PersistentEntity;
import java.util.Arrays;
import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.springframework.stereotype.Component;

@Aspect
@Component
public class ModifierAspect {

  @Before("execution(* com.github.jactor.persistence.repository.*Repository.save(..))")
  public Object modifyPersistentEntity(JoinPoint joinPoint) {
    Arrays.stream(joinPoint.getArgs())
        .filter(obj -> obj instanceof PersistentEntity)
        .map(obj -> (PersistentEntity) obj)
        .filter(persistentEntity -> persistentEntity.getId() != null)
        .forEach(PersistentEntity::modify);

    return null;
  }
}

File: src/main/kotlin/com/github/jactor/persistence/aop/ModifierAspect.kt
package com.github.jactor.persistence.aop

import com.github.jactor.persistence.entity.PersistentEntity
import org.aspectj.lang.JoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.springframework.stereotype.Component
import java.util.Arrays

@Aspect
@Component
class ModifierAspect {
    @Before("execution(* com.github.jactor.persistence.repository.*Repository.save(..))")
    fun modifyPersistentEntity(joinPoint: JoinPoint): Any? {
        Arrays.stream(joinPoint.args)
            .filter { obj: Any? -> obj is PersistentEntity<*> }
            .map { obj: Any -> obj as PersistentEntity<*> }
            .filter { persistentEntity: PersistentEntity<*> -> persistentEntity.id != null }
            .forEach { obj: PersistentEntity<*> -> obj.modify() }
        return null
    }
}